pkg/tagset/slice_builder.go

```diff
@@ -0,0 +1,197 @@
+//
+// The Add* methods may only be called in the "adding tags" stage. No methods
+// may be called in the "closed" stage.
+//
+// In general, the easiest way to ensure these stages are followed (and allow reviewers
+// to verify this) is to use a builder in a single method, as shown in the example.  Avoid
+// storing builders in structs.
+//
+// SliceBuilders are not threadsafe.
+type SliceBuilder struct {
+	factory Factory
+
+	// Number of levels this builder supports
+	levels int
+
+	// Offsets into the tags/hashes arrays, such that level L is represented as
+	// tags[offsets[L]:offsets[L+1]]. This slice has length (levels+1).
+	offsets []int
+
+	// The built slice of tags
+	tags []string
+
+	// Hashes of individual tags, with each element corresponding to the hash
+	// of that tag. INVARIANT: len(tags) == len(hashes), cap(tags) ==
+	// cap(hashes)
+	hashes []uint64
+
+	// Hashes of all tags at each level. This slice has length levels.
+	levelHashes []uint64
+
+	// all seen tags (regardless of level)
+	seen map[uint64]struct{}
+}
+
+// newSliceBuilder creates a new builder. This must be reset()
+// before use.
+func newSliceBuilder(factory Factory) *SliceBuilder {
+	return &SliceBuilder{
+		factory: factory,
+	}
+}
+
+// reset the builder, preparing it for re-use
+func (bldr *SliceBuilder) reset(levels, capacity int) {
+	bldr.levels = levels
+
+	// expand bldr.offsets to at least length levels+1
+	if bldr.offsets == nil || cap(bldr.offsets) < levels+1 {
+		bldr.offsets = make([]int, levels+1)
+	} else {
+		bldr.offsets = bldr.offsets[:levels+1]
+	}
+
+	for i := 0; i < levels+1; i++ {
+		bldr.offsets[i] = 0
+	}
+
+	// expand bldr.offsets to at least length levels
+	if bldr.levelHashes == nil || cap(bldr.levelHashes) < levels {
+		bldr.levelHashes = make([]uint64, levels)
+	} else {
+		bldr.levelHashes = bldr.levelHashes[:levels]
+	}
+
+	for i := 0; i < levels; i++ {
+		bldr.levelHashes[i] = 0
+	}
+
+	bldr.tags = make([]string, 0, capacity)
+	bldr.hashes = make([]uint64, 0, capacity)
+	bldr.seen = map[uint64]struct{}{}
+}
+
+// Add adds the given tag to the builder at the given level. If the tag is
+// already in the builder, it is not added again (regardless of the existing
+// tag's level).
+func (bldr *SliceBuilder) Add(level int, tag string) {
+	h := murmur3.StringSum64(tag)
+	if _, seen := bldr.seen[h]; seen {
+		return
+	}
+	bldr.seen[h] = struct{}{}
+
+	newLen := len(bldr.tags) + 1
+	insertAt := bldr.offsets[level+1]
+
+	var newTags []string
+	var newHashes []uint64
+
+	// reallocate the storage if there is not enough room in the existing
+	// array, and copy the existing data into the new array
+	if cap(bldr.tags) < newLen {
+		newTags = make([]string, newLen, newLen*2)
+		newHashes = make([]uint64, newLen, newLen*2)
+
+		copy(newTags[:insertAt], bldr.tags[:insertAt])
+		copy(newTags[insertAt+1:], bldr.tags[insertAt:])
+
+		copy(newHashes[:insertAt], bldr.hashes[:insertAt])
+		copy(newHashes[insertAt+1:], bldr.hashes[insertAt:])
```

Comment:
While I fully acknowledge that this `Add` implementation is highly optimized and very clean (subjectively speaking from my pov), I'm doubtful that this code will be very forward-maintainable due to the high amount of low-level work with offset/item shifting, flat tag array co-mingled levels, and exponential resizing all in the same file. I would personally advocate at least for careful decomposition here into logical functions even if the price of additional function calls will reduce performance to ensure that we can easily maintain it in the future.

Reply:
Done!